lib/utils/remote-build.ts

```diff
@@ -0,0 +1,213 @@
+			cancelBuildIfNecessary(build).then(() => {
+				stream.end();
+				process.exit(0);
+			});
+		});
+
+		stream.on('data', getBuilderMessageHandler(build));
+		stream.on('end', resolve);
+		stream.on('error', reject);
+	}).then(() => {
+		build.running = false;
+	});
+}
+
+async function handleBuilderMetadata(obj: BuilderMessage, build: RemoteBuild) {
+	switch (obj.resource) {
+		case 'cursor':
+			const readline = await import('readline');
+
+			if (obj.value == null) {
+				return;
+			}
+
+			const match = obj.value.match(CURSOR_METADATA_REGEX);
+
+			if (!match) {
+				// FIXME: Make this error nicer.
+				console.log(
+					'Warning: ignoring unknown builder output. You may experience odd build output. ' +
+						'Maybe you need to update resin-cli?',
+				);
+				return;
+			}
+
+			const value = match[1];
+			const amount = match[2] || 1;
+
+			switch (value) {
+				case 'erase':
+					readline.clearLine(process.stdout, 0);
+					process.stdout.write('\r');
+					break;
+				case 'up':
+					readline.moveCursor(process.stdout, 0, -amount);
+					break;
+				case 'down':
+					readline.moveCursor(process.stdout, 0, amount);
+					break;
+			}
+
+			break;
+		case 'buildLogId':
+			// The name of this resource is slightly dated, but this is the release
+			// id from the API. We need to save this so that if the user ctrl+c's the
+			// build we can cancel it on the API.
+			build.releaseId = parseInt(obj.value!, 10);
+			break;
+	}
+}
+
+function getBuilderMessageHandler(
+	build: RemoteBuild,
+): (obj: BuilderMessage) => Promise<void> {
+	return async (obj: BuilderMessage) => {
+		if (DEBUG_MODE) {
+			console.log(`[debug] handling message: ${JSON.stringify(obj)}`);
+		}
+		if (obj.type != null && obj.type === 'metadata') {
+			return handleBuilderMetadata(obj, build);
+		}
+		if (obj.message) {
+			const readline = await import('readline');
+			readline.clearLine(process.stdout, 0);
+
+			if (!SUPPRESS_OUTPUT) {
+				const message = obj.message.replace(TRIM_REGEX, '');
+				if (obj.replace) {
+					process.stdout.write(`\r${message}`);
+				} else {
+					console.log(message);
+				}
+			}
+		}
+	};
+}
+
+async function cancelBuildIfNecessary(build: RemoteBuild): Promise<void> {
+	if (build.running && build.releaseId != null) {
+		await build.sdk.pine.patch({
+			resource: 'release',
+			id: build.releaseId,
+			body: {
+				status: 'cancelled',
+				end_timestamp: Date.now(),
+			},
+		});
+	}
+}
+
+async function getRequestStream(build: RemoteBuild): Promise<Stream.Duplex> {
+	// Tar the directory so that we can send it to the builder
+	const tarStream = await tarDirectory(build.source);
+
+	if (DEBUG_MODE) {
+		console.log('[debug] Opening builder connection');
+	}
+	const post = request.post({
+		url: await getBuilderEndpoint(build.baseUrl, build.owner, build.app),
+		auth: {
+			bearer: build.auth,
+		},
+	});
+	if (DEBUG_MODE) {
+		console.log('[debug] Opened builder connection. Streaming data...');
+	}
+
+	tarStream.pipe(post);
```

Comment:
Right now, there's no output until the context directory is tarred and fully uploaded. That's not great if you have a large local directory. There was quite a bit of confusion from local push where people didn't understand what it was doing and why it took so long, because they accidentally had enormous local folders they were pushing over the network each time.

Can we add some initial output and a nice spinner while we do the upload?